Protect object creation from dropping referenced objects

whitehawk opened the pull request:
Protect object creation from dropping referenced objects

Problem description:
If an object had been created inside a transaction and it had a dependency on 
some other object (like schema, custom data type, etc), and that referenced 
object had been dropped from another transaction before the first transaction 
had been committed, the new object was broken. 

Ex. steps:
```
Session 1: create type test_type as enum ('one', 'two');
Session 1: begin;
Session 1: create table test_table(a test_type);
Session 2: drop type test_type;
Session 1: commit;
Session 1: select * from test_table;
```
gave result:
```
ERROR:  cache lookup failed for type 16385 (lsyscache.c:2796)
```

The issue reproduced for all dependencies that are registered in 'pg_depend'.

Root cause:
No protection from stealing the objects that are marked as dependencies in
'pg_depend' before the transaction is committed.

Fix:
Lock the referenced object with AccessShareLock when the dependency is created.
Lock is done only for 'normal' dependencies (which means here 'normal
relationship between separately-created objects', when the dependent object may
be dropped without affecting the referenced object, while the referenced object
may only be dropped with CASCADE, in which case the dependent object is dropped
too) and 'auto' (meaning that the dependent object can be dropped separately
from the referenced object, and should be automatically dropped if the
referenced object is dropped).
After the lock is released, do a recheck of the object's presence and error out
if the object was dropped while we waited for the lock.

whitehawk committed:
Preliminary fix

whitehawk committed:
Add tests

whitehawk committed:
Merge branch 'adb-6.x-dev' into ADBDEV-7586

whitehawk committed:
Fix tests

whitehawk committed:
Add comments

whitehawk committed:
Update comments in the test

hilltracer commented on src/test/isolation2/expected/dependency.out:
```diff
@@ -0,0 +1,408 @@
+2: commit;
+COMMIT
+1<:  <... completed>
+ERROR:  function test5_default_value_function() does not exist  (seg0 127.0.1.1:6002 pid=2718914)
```
Do we have matchsubs for it? Maybe also use XXXXX? The same questions about other strings. (See 307, 355, 407 lines)

whitehawk replied:
There is no need to make dedicated matchsubs for seg connection info - it is already handled by `atmsort.pm` + init files.

RekGRpth commented:
> Deadlocks are possible. But it can be resolved by the deadlock detection algorithm. So, these 2 transactions trigger ERROR:  deadlock detected with one of the transactions aborted.

Could you add a test for this?

RekGRpth commented on src/test/isolation2/expected/dependency.out:
```diff
@@ -0,0 +1,408 @@
+2: commit;
+COMMIT
+1<:  <... completed>
+ERROR:  schema "test_1_schema" does not exist  (seg0 127.0.1.1:6002 pid=2715327)
```
Why did the error occur on the segment and not on the coordinator?

whitehawk replied:
Because in this case after the lock is released coordinator doesn't check if the object is still present or not, so the error is caught on the segment. Possibly for GPDB we could live without re-check on the coordinator. But after some thinking I've decided to add a re-check of the object presence (like it is already done in [`shdepLockAndCheckObject`](https://github.com/arenadata/gpdb/blob/b411aec5f1bd941f619f21b98d80c8206ef300ec/src/backend/catalog/pg_shdepend.c#L982)). So now the error occurs on the coordinator always. It aligns the behavior with standalone Postgres if this patch is applied to it. Thus we will be able to propose it 'as is' (at least the part not related to tests).

RekGRpth commented:
Is it expected that the patch does not cover the case where the type being deleted is used in the function body?
<details><summary>Diff</summary>
<p>

```diff
diff --git a/src/test/isolation2/expected/dependency.out b/src/test/isolation2/expected/dependency.out
index f7a88622630..5d6df295a02 100644
--- a/src/test/isolation2/expected/dependency.out
+++ b/src/test/isolation2/expected/dependency.out
@@ -121,26 +121,28 @@ CREATE
 
 1: begin;
 BEGIN
-1: create function test_3_function(a test_3_type) returns text as $$ select 'Return ' || a; /**/ $$ language sql;
+1: create function test_3_function(a text) returns text as $$ select 'Return ' || a::test_3_type; /**/ $$ language sql;
 CREATE
 
 2&: drop type test_3_type;  <waiting ...>
+FAILED:  Forked command is not blocking; got output: DROP
 
 1: commit;
 COMMIT
 
 2<:  <... completed>
-ERROR:  cannot drop type test_3_type because other objects depend on it
-DETAIL:  function test_3_function(test_3_type) depends on type test_3_type
-HINT:  Use DROP ... CASCADE to drop the dependent objects too.
+FAILED:  Execution failed
 
 1: select test_3_function('one');
- test_3_function 
------------------
- Return one      
-(1 row)
+ERROR:  type "test_3_type" does not exist
+LINE 2:     select 'Return ' || a::test_3_type; /**/
+                                   ^
+QUERY:  
+    select 'Return ' || a::test_3_type; /**/
 
-drop type test_3_type cascade;
+CONTEXT:  SQL function "test_3_function" during inlining
+
+drop function test_3_function(a text) cascade;
 DROP
 
 -- Check if dependency is dropped before the creation of the dependent object.
@@ -152,13 +154,21 @@ BEGIN
 BEGIN
 2: drop type test_3_type;
 DROP
-1&: create function test_3_function(a test_3_type) returns text as $$ select 'Return ' || a; /**/ $$ language sql;  <waiting ...>
+1&: create function test_3_function(a text) returns text as $$ select 'Return ' || a::test_3_type; /**/ $$ language sql;  <waiting ...>
+FAILED:  Forked command is not blocking; got output: CREATE
 
 2: commit;
 COMMIT
 1<:  <... completed>
-ERROR:  cache lookup failed for type XXXXX (lsyscache.c:XXX)
-CONTEXT:  SQL function "test_3_function"
+FAILED:  Execution failed
+1: select test_3_function('one');
+ERROR:  type "test_3_type" does not exist
+LINE 2:     select 'Return ' || a::test_3_type; /**/
+                                   ^
+QUERY:  
+    select 'Return ' || a::test_3_type; /**/
+
+CONTEXT:  SQL function "test_3_function" during inlining
 1: end;
 END
 
diff --git a/src/test/isolation2/sql/dependency.sql b/src/test/isolation2/sql/dependency.sql
index b8bfe65a214..fe3afe39140 100644
--- a/src/test/isolation2/sql/dependency.sql
+++ b/src/test/isolation2/sql/dependency.sql
@@ -84,8 +84,8 @@ $$ language sql;
 create type test_3_type as enum ('one', 'two');
 
 1: begin;
-1: create function test_3_function(a test_3_type) returns text as $$
-    select 'Return ' || a; /**/
+1: create function test_3_function(a text) returns text as $$
+    select 'Return ' || a::test_3_type; /**/
 $$ language sql;
 
 2&: drop type test_3_type;
@@ -96,19 +96,20 @@ $$ language sql;
 
 1: select test_3_function('one');
 
-drop type test_3_type cascade;
+drop function test_3_function(a text) cascade;
 
 -- Check if dependency is dropped before the creation of the dependent object.
 create type test_3_type as enum ('one', 'two');
 1: begin;
 2: begin;
 2: drop type test_3_type;
-1&: create function test_3_function(a test_3_type) returns text as $$
-    select 'Return ' || a; /**/
+1&: create function test_3_function(a text) returns text as $$
+    select 'Return ' || a::test_3_type; /**/
 $$ language sql;
 
 2: commit;
 1<:
+1: select test_3_function('one');
 1: end;
 
 -- Case 4. Function dependency on the language.
```
</p>
</details>

RekGRpth commented:
Could you benchmark the performance degradation due to adding new extra locks?

hilltracer reviewed (commented):
Should we add the ticket number into the PR description?

Maybe add a regression test that checks pg_locks to ensure we don't add extra locks for pinned objects (e.g., int4, text, etc.).

Maybe also add a test that checks pg_locks to show how the added locks are distributed across our MPP between the coordinator and segments?

whitehawk committed:
Check after lock if dependency was concurrently dropped

whitehawk committed:
Add text parser dep lock support + test case

whitehawk committed:
Add text search template dep lock support + test case

whitehawk committed:
Add fdw dep lock support + test case

whitehawk committed:
Add server dep lock support + test case

whitehawk committed:
Add protocol dep lock support + test case

whitehawk committed:
Add ts dictionary dep lock support + test case

whitehawk committed:
Add operator dep lock support + test case

whitehawk committed:
Add operator class dep lock support + test case

whitehawk committed:
Refactor

whitehawk committed:
Handle non-existing type when defining I/O procs

whitehawk committed:
Add operator family dep lock support + test case

whitehawk committed:
Build fix

whitehawk committed:
Update test

whitehawk committed:
Refactor

whitehawk committed:
Add lock count check into first couple of test cases

whitehawk committed:
Add deadlock test case

whitehawk committed:
Add view test

whitehawk committed:
Add materialized view test

whitehawk committed:
Add alter column test

whitehawk committed:
Add rule test

whitehawk commented:
@hilltracer 

> Should we add the ticket number into the PR description?

I usually do not add it, as it can be obtained from the branch name. And it is not mandated by the review checklist.

> Maybe add a regression test that checks pg_locks to ensure we don't add extra locks for pinned objects (e.g., int4, text, etc.).

> Maybe also add a test that checks pg_locks to show how the added locks are distributed across our MPP between the coordinator and segments?

I've added check of pg_locks into the first couple of test cases.

---------------

@RekGRpth 

> Could you benchmark the performance degradation due to adding new extra locks?

I've done performance testing with the help of functions below:
```
CREATE OR REPLACE FUNCTION create_many_objects()
RETURNS VOID
AS $$
BEGIN
	for counter in 1..1000 loop
		EXECUTE FORMAT($fmt$CREATE TYPE new_type_%1$s as enum ('one', 'two');$fmt$,
			counter);
		EXECUTE FORMAT($fmt$CREATE TABLE new_table_%1$s (a new_type_%1$s, b new_type_%1$s, c new_type_%1$s) DISTRIBUTED BY (a);$fmt$,
			counter);
	end loop;
END$$
LANGUAGE plpgsql VOLATILE
EXECUTE ON MASTER;

CREATE OR REPLACE FUNCTION remove_many_objects()
RETURNS VOID
AS $$
BEGIN
	-- cleanup
	for counter in 1..1000 loop
		EXECUTE FORMAT($fmt$DROP table new_table_%1$s;$fmt$,
			counter);
		EXECUTE FORMAT($fmt$DROP TYPE new_type_%1$s;$fmt$,
			counter);
	end loop;
END$$
LANGUAGE plpgsql VOLATILE
EXECUTE ON MASTER;
```

commands used for perf measurement:
```
for i in {1..10};
do
	TIME="\t%E sec" time psql postgres -c "select create_many_objects();" > /dev/null
	psql postgres -c "select remove_many_objects();" > /dev/null
done
```

results without the patch:
```
        0:13.64 sec
        0:19.30 sec
        0:22.13 sec
        0:21.60 sec
        0:19.96 sec
        0:20.14 sec
        0:21.98 sec
        0:21.73 sec
        0:22.72 sec
        0:21.62 sec
```
total 204,82 sec.

results with the patch:
```
        0:14.24 sec
        0:21.03 sec
        0:22.06 sec
        0:22.06 sec
        0:20.36 sec
        0:20.16 sec
        0:21.23 sec
        0:21.71 sec
        0:21.60 sec
        0:20.81 sec

```
total 205,26 sec.

So the difference between total numbers looks neglectable. 

> Could you add tests for creating a view that depends on the function being removed, 
> and for changing the column type of a table, and also for the rules that depend on the table being dropped?

Added

> Is it expected that the patch does not cover the case where the object being deleted is used in the function body?

Yes, according to [postgres documentation](https://www.postgresql.org/docs/current/ddl-depend.html):
_For a user-defined function or procedure whose body is defined as a string literal, PostgreSQL tracks dependencies associated with the function's externally-visible properties, such as its argument and result types, but not dependencies that could only be known by examining the function body._

>> Deadlocks are possible. But it can be resolved by the deadlock detection algorithm. So, these 2 transactions trigger ERROR: deadlock detected with one of the transactions aborted.

> Could you add a test for this?

Added.

hilltracer commented on src/backend/catalog/pg_depend.c:
```diff
@@ -787,3 +797,108 @@ get_index_constraint(Oid indexId)
+		if ((cacheId == TYPEOID || cacheId == OPFAMILYOID) &&
+			!SearchSysCacheExists1(cacheId,
+								   ObjectIdGetDatum(referenced->objectId)))
+			return;
```
I didn't find test for it. Where is it tested?

whitehawk replied:
for case `cacheId == TYPEOID` - you will not be able even to start up the cluster without this line.
for case `cacheId == OPFAMILYOID` - without it test `Case 16. Index dependency on the operator class` will fail,

hilltracer commented on src/backend/catalog/pg_depend.c:
```diff
@@ -91,6 +94,13 @@ recordMultipleDependencies(const ObjectAddress *depender,
+			 * Lock the referenced object to protect it from dropping by
+			 * another transaction.
+			 */
+			if (behavior == DEPENDENCY_NORMAL || behavior == DEPENDENCY_AUTO)
```
Do you have test for DEPENDENCY_AUTO?

Is it expected?
```sql
-- current session:
gpadmin=# create sequence test_22_seq;
CREATE SEQUENCE
gpadmin=# begin;
BEGIN
gpadmin=# create table test_22_table(id int default nextval('test_22_seq'));
CREATE TABLE

-- other session: 
drop sequence test_22_seq; 

-- current session:
gpadmin=# commit;
COMMIT
gpadmin=# insert into test_22_table default values;
ERROR:  relation not found (OID 19102)  (entry db 172.18.0.2:6000 pid=54292)
DETAIL:  This can be validly caused by a concurrent delete operation on this object.
gpadmin=# select * from test_22_table;
 id 
----
(0 rows)
```

whitehawk replied:
> Do you have test for DEPENDENCY_AUTO?

Yes, it is `Case 17`.

> Is it expected?

I've fixed handling of the dependency on a sequence and added this test.

hilltracer reviewed (commented):
> So the difference between total numbers looks neglectable.

Good result, but it only shows performance in a single-session scenario. What happens if multiple sessions create objects in parallel and all access the same referenced objects, putting more pressure on the lock manager?
Maybe run another experiment where many clients each execute `create_many_objects()` in their own transaction?

Also maybe discover Tom Lane’s concern about blowing past `max_locks_per_transaction` during a big `pg_restore --single-transaction`. So we can do something like `SELECT count(*) FROM pg_locks WHERE pid = pg_backend_pid();`  after `create_many_objects()` but before commit, query. And compare counts pre/post patch.

Should we also check behavior with `REPEATABLE READ` and `SERIALIZABLE` in tests? Maybe add one case to show behavior with other isolation level, I hope it's should be identical.

whitehawk committed:
Handle dependency on a sequence + test for it

whitehawk committed:
Add test for REPEATABLE READ isolation level

whitehawk committed:
Merge branch 'adb-6.x-dev' into ADBDEV-7586

whitehawk committed:
Limit dep lock only for sequences

hilltracer reviewed (dismissed):
Thanks for the detailed experiments. I think we're now ready to present our results for Tom Lane and co.

RekGRpth commented on src/backend/catalog/pg_depend.c:
```diff
@@ -787,3 +797,121 @@ get_index_constraint(Oid indexId)
+	}
+
+	if (cacheId >= 0)
+	{
```
```suggestion
				break;
			}
			return;
		default:
			return;
	}

	Assert(cacheId >= 0);
```

whitehawk committed:
Refactor

RekGRpth commented on src/backend/catalog/pg_depend.c:
```diff
@@ -787,3 +797,119 @@ get_index_constraint(Oid indexId)
+static void
+depLockAndCheckObject(const ObjectAddress *referenced)
+{
+	int			cacheId = -1;
+	char	   *objName;
```
may be
```suggestion
	enum SysCacheIdentifier cacheId;
	const char *objName;
```
and remove assert below?

whitehawk committed:
Some minor refactoring

RekGRpth commented on src/test/isolation2/sql/dependency.sql:
```diff
@@ -0,0 +1,801 @@
+1:  with cte as (select * from pg_locks where pid = pg_backend_pid() and locktype = 'object')
+select count(1), gp_segment_id from pg_locks
+where classid in (select classid from cte) and objid in (select objid from cte)
+group by gp_segment_id order by gp_segment_id;
```
too hard!
why not simply
```suggestion
1: select l.gp_segment_id, count(1)
from pg_locks l
join pg_locks c using (locktype, classid, objid)
where c.pid = pg_backend_pid() and c.locktype = 'object'
group by 1 order by 1;
```

whitehawk committed:
Update test

RekGRpth commented on src/backend/catalog/pg_depend.c:
```diff
@@ -787,3 +797,117 @@ get_index_constraint(Oid indexId)
+	{
+		case OCLASS_SCHEMA:
+			cacheId = NAMESPACEOID;
+			objName = "namespace";
```
why not
```suggestion
			objType = "namespace";
			objName = get_namespace_name(referenced->objectId);
```
for a more understandable error message?

whitehawk replied:
I do not want to add not-100%-necessary code, like this addional request to sys cache. The problem itself seems pretty rear. I consider current error message info is enough to provide good hints for the user.

RekGRpth replied:
```sql
create function schema.function(a1 schema1.type, a2 schema2.type, a3 schema3.type, a4 schema4.type, a5 schema5.type, a6 schema6.type, a7 schema7.type, a8 schema8.type) returns setof schema9.type as $$ select i from generate_series(1,5)i; $$ language sql;
ERROR:  namespace 33367 was concurrently dropped
```
which of 10 namespaces?

RekGRpth replied:
> this addional request to sys cache

This can only be done before an error message is issued.

whitehawk replied:
Once again, I think that overloading with more code may be a blocker when later we will try to push this change to Postgres upstream. As for GPDB, performance of the DDL commands may be not a big deal, while for upstream Postgres that most likely will be an issue. 

> This can only be done before an error message is issued.

That can be deduced by knowing which objects are still present, for ex by smth like:
```
select * from pg_namespace where nspname in ('schema1', 'schema2', 'schema3', 'schema4', 'schema5', 'schema6', 'schema7', 'schema8', 'schema9');
```

RekGRpth commented on src/test/isolation2/sql/dependency.sql:
```diff
@@ -0,0 +1,803 @@
+1<:
+1: end;
+
+-- Case 11. Server dependency on the foreigh data wrapper.
```
```suggestion
-- Case 11. Server dependency on the foreign data wrapper.
```

whitehawk committed:
Fix typo

RekGRpth commented on src/test/isolation2/sql/dependency.sql:
```diff
@@ -0,0 +1,803 @@
+
+2<:
+
+! psql isolation2test -c '\deu';
```
why not simply
```suggestion
SELECT srvname, usename FROM pg_user_mappings ORDER BY 1, 2;
```

RekGRpth commented on src/test/isolation2/sql/dependency.sql:
```diff
@@ -0,0 +1,803 @@
+
+2<:
+
+1: select * from test_16_table where a = 0;
```
What does this line check?

whitehawk replied:
It is required to error out if the patch didn't work:
```
1: select * from test_16_table where a = 0;
ERROR:  cache lookup failed for opclass 16518 (lsyscache.c:1212)
```

whitehawk committed:
Update test

RekGRpth commented on src/test/isolation2/expected/dependency.out:
```diff
@@ -0,0 +1,1200 @@
+2: commit;
+COMMIT
+1<:  <... completed>
+ERROR:  relation "test_21_table_2" does not exist
+LINE 1: ... insert to test_21_table_1 do instead insert into test_21_ta...
```
Why not
```sql
ERROR:  table ##### was concurrently dropped
```

whitehawk replied:
Because rules on a table were protected before my patch, but I've added this test case as you've requested it. So here the error message is different.

RekGRpth replied:
Is it possible to make the rest of the error messages like this?

whitehawk replied:
Currently the error message is aligned with similar for shared dependencies in [shdepLockAndCheckObject()](https://github.com/arenadata/gpdb/blob/915aeb3eb82371b62b3ac3724cc84230b43d4671/src/backend/catalog/pg_shdepend.c#L982), so I think that current variant is better.
